smart-contract-extensions/contracts/hooks/DiscountCodeHook.sol

```diff
@@ -0,0 +1,81 @@
+pragma solidity 0.5.17;
+
+import '@openzeppelin/contracts/math/SafeMath.sol';
+import '@openzeppelin/contracts/cryptography/ECDSA.sol';
+import '@openzeppelin/contracts/access/roles/WhitelistAdminRole.sol';
+import 'unlock-abi-7/ILockKeyPurchaseHookV7.sol';
+import 'unlock-abi-7/IPublicLockV7.sol';
+
+/**
+ * @notice Used with a Lock to offer discounts if the user enters a code.
+ */
+contract DiscountCodeHook is ILockKeyPurchaseHookV7, WhitelistAdminRole
+{
+  using SafeMath for uint;
+
+  /**
+   * @notice The code expressed as an address where the private key is
+   * keccak256(abi.encode(code, lock.address)).
```

Comment:
Are you able to quantify how secure this is? (ie: many bits of entropy or other metric)
Asking because, while this seems pretty good intuitively, I actually don't know.